File: my_cpp_pkg/src/number_publisher.cpp
#include <rclcpp/rclcpp.hpp>
#include <example_interfaces/msg/int64.hpp>

using namespace std::placeholders;

class NumberPublisherNode : public rclcpp::Node
{
public:
  NumberPublisherNode()
      : Node("number_publisher")
  {
    this->declare_parameter("number", 9);
    this->declare_parameter("timer_period", 1.0);

    number_ = this->get_parameter("number").as_int();
    timerPeriod_ = this->get_parameter("timer_period").as_double();

    pCallbackParams_ = this->add_post_set_parameters_callback(
        std::bind(&NumberPublisherNode::callbackParameters, this, _1));

    pPublisher_ = this->create_publisher<example_interfaces::msg::Int64>("number", 10);
    timer_ = this->create_wall_timer(
        std::chrono::duration<double>(timerPeriod_),
        std::bind(&NumberPublisherNode::publishNumber, this));

    RCLCPP_INFO(this->get_logger(), "Number publisher started.");
  }

private:
  void publishNumber()
  {
    auto msg = example_interfaces::msg::Int64();
    msg.data = number_;
    pPublisher_->publish(msg);
  }

  void callbackParameters(const std::vector<rclcpp::Parameter> &params)
  {
    for (const auto &param : params)
    {
      if (param.get_name() == "number")
      {
        number_ = param.as_int();
      }
      else if (param.get_name() == "timer_period")
      {
        timerPeriod_ = param.as_double();
        timer_->cancel();
        timer_ = this->create_wall_timer(
            std::chrono::duration<double>(timerPeriod_),
            std::bind(&NumberPublisherNode::publishNumber, this));
      }
    }
  }

  int64_t number_;
  double timerPeriod_;

  rclcpp::Publisher<example_interfaces::msg::Int64>::SharedPtr pPublisher_;
  rclcpp::TimerBase::SharedPtr timer_;

  PostSetParametersCallbackHandle::SharedPtr pCallbackParams_;
};

int main(int argc, char **argv)
{
  rclcpp::init(argc, argv);

  auto pNode = std::make_shared<NumberPublisherNode>();
  rclcpp::spin(pNode);

  rclcpp::shutdown();
  return 0;
}

File: my_py_pkg/my_py_pkg/number_publisher.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from rclpy.parameter import Parameter
from example_interfaces.msg import Int64


class NumberPublisherNode(Node):
    def __init__(self):
        super().__init__("number_publisher")

        self.declare_parameter(name="number", value=9)
        self.declare_parameter(name="timer_period", value=1.0)

        self.number_ = self.get_parameter(name="number").value
        self.timer_period_ = self.get_parameter(name="timer_period").value

        self.add_post_set_parameters_callback(self.callback_parameters)

        self.publisher_ = self.create_publisher(
            msg_type=Int64, topic="number", qos_profile=10
        )
        self.timer_ = self.create_timer(
            timer_period_sec=self.timer_period_, callback=self.publish_number
        )

        self.get_logger().info("Number Publisher started.")

    def publish_number(self):
        msg = Int64()
        msg.data = self.number_
        self.publisher_.publish(msg)

    def callback_parameters(self, params: list[Parameter]):
        for param in params:
            if param.name == "number":
                self.number_ = param.value
            if param.name == "timer_period":
                self.timer_period_ = param.value
                self.timer_.destroy()
                self.timer_ = self.create_timer(
                    timer_period_sec=self.timer_period_, callback=self.publish_number
                )


def main(args=None):
    rclpy.init(args=args)

    node = NumberPublisherNode()
    rclpy.spin(node=node)

    rclpy.shutdown()


if __name__ == "__main__":
    main()
